pkg/resource/typed/types_resource_test.go

```diff
@@ -0,0 +1,78 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+package typed_test
+
+import (
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+
+	"github.com/cosi-project/runtime/pkg/resource"
+	"github.com/cosi-project/runtime/pkg/resource/meta"
+	"github.com/cosi-project/runtime/pkg/resource/typed"
+)
+
+type TestSpec struct {
+	Var int
+}
+
+func (t TestSpec) DeepCopy() TestSpec {
+	return t
+}
+
+type Test = typed.TypedResource[TestSpec, TestRD]
+
+func NewTest(md resource.Metadata, spec TestSpec) *Test {
+	return typed.NewTypedResource[TestSpec, TestRD](md, spec)
+}
+
+type TestRD struct{}
+
+func (TestRD) ResourceDefinition(md resource.Metadata, spec TestSpec) meta.ResourceDefinitionSpec {
+	return meta.ResourceDefinitionSpec{
+		DisplayType: "test definition",
+	}
+}
+
+// ireturn thinks that we return interface here, but we do not
+func must[T any](t T, err error) T { //nolint:ireturn
+	if err != nil {
+		panic(err)
+	}
+
+	return t
+}
+
+// checks that both parameterers implement resource.Resource and has equal specs
+func checkResourceEqualSpec(t *testing.T, a, b resource.Resource) {
+	assert.Equal(t, a.Spec(), b.Spec())
+}
+
+func TestTypedResource(t *testing.T) {
+	t.Parallel()
+	asrt := assert.New(t)
+	md := resource.NewMetadata("default", "type", "aaa", resource.VersionUndefined)
+	spec := TestSpec{
+		Var: 42,
+	}
+
+	res := NewTest(md, spec)
+
+	// check that stored metadata is actually valid
+	asrt.Equal(res.Metadata().ID(), "aaa")
+	asrt.Equal(res.Metadata().Version(), must(resource.ParseVersion("1")))
+
+	// check that deep copy doesn't modify metadata or spec data
+	resCopy := res.DeepCopy()
+	asrt.True(resCopy.Metadata().Equal(*res.Metadata()))
+	checkResourceEqualSpec(t, resCopy, res)
+
+	// check that modifying one doesn't modify the other
+	res.TypedSpec().Var = 45
+	asrt.NotEqual(resCopy.Spec(), res.Spec())
+
+	// check that getting resource definition actually works on phantom types
+	asrt.Equal(res.ResourceDefinition().DisplayType, "test definition")
```

Comment:
I think we need `asrt.Implements` that `*Test` implements a `resource.Resource`

Reply:
Done using:
```
var _ resource.Resource = (*Test)(nil)
```